nvflare/apis/impl/wf_comm_server.py

```diff
@@ -0,0 +1,1012 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import threading
+import time
+from threading import Lock
+from typing import List, Optional, Tuple, Union
+
+from nvflare.apis.client import Client
+from nvflare.apis.controller_spec import ClientTask, SendOrder, Task, TaskCompletionStatus
+from nvflare.apis.fl_constant import FLContextKey
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.job_def import job_from_meta
+from nvflare.apis.responder import Responder
+from nvflare.apis.shareable import ReservedHeaderKey, Shareable, make_copy
+from nvflare.apis.signal import Signal
+from nvflare.apis.wf_comm_spec import WFCommSpec
+from nvflare.fuel.utils.config_service import ConfigService
+from nvflare.security.logging import secure_format_exception
+from nvflare.widgets.info_collector import GroupInfoCollector, InfoCollector
+
+from .any_relay_manager import AnyRelayTaskManager
+from .bcast_manager import BcastForeverTaskManager, BcastTaskManager
+from .send_manager import SendTaskManager
+from .seq_relay_manager import SequentialRelayTaskManager
+from .task_manager import TaskCheckStatus, TaskManager
+
+_TASK_KEY_ENGINE = "___engine"
+_TASK_KEY_MANAGER = "___mgr"
+_TASK_KEY_DONE = "___done"
+
+# wait this long since client death report before treating the client as dead
+_CONFIG_VAR_DEAD_CLIENT_GRACE_PERIOD = "dead_client_grace_period"
+
+# wait this long since job schedule time before starting to check dead clients
+_CONFIG_VAR_DEAD_CLIENT_CHECK_LEAD_TIME = "dead_client_check_lead_time"
+
+
+def _check_positive_int(name, value):
+    if not isinstance(value, int):
+        raise TypeError("{} must be an instance of int, but got {}.".format(name, type(name)))
+    if value < 0:
+        raise ValueError("{} must >= 0.".format(name))
+
+
+def _check_inputs(task: Task, fl_ctx: FLContext, targets: Union[List[Client], List[str], None]):
+    if not isinstance(task, Task):
+        raise TypeError("task must be an instance of Task, but got {}".format(type(task)))
+
+    if not isinstance(fl_ctx, FLContext):
+        raise TypeError("fl_ctx must be an instance of FLContext, but got {}".format(type(fl_ctx)))
+
+    if targets is not None:
+        if not isinstance(targets, list):
+            raise TypeError("targets must be a list of Client or string, but got {}".format(type(targets)))
+
+        for t in targets:
+            if not isinstance(t, (Client, str)):
+                raise TypeError(
+                    "targets must be a list of Client or string, but got element of type {}".format(type(t))
+                )
+
+
+def _get_client_task(target, task: Task):
+    for ct in task.client_tasks:
+        if target == ct.client.name:
+            return ct
+    return None
+
+
+class WFCommServer(Responder, WFCommSpec):
```

Comment:
I know this is taken from previous.
Do we still need a Responder class/abstract?